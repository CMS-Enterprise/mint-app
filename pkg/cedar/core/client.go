package cedarcore

import (
	"context"
	"net/http"
	"time"

	ld "gopkg.in/launchdarkly/go-server-sdk.v5"

	"github.com/go-openapi/runtime"
	httptransport "github.com/go-openapi/runtime/client"
	"github.com/go-openapi/strfmt"
	cache "github.com/patrickmn/go-cache"
	"go.uber.org/zap"

	"github.com/cmsgov/easi-app/pkg/appcontext"
	apiclient "github.com/cmsgov/easi-app/pkg/cedar/core/gen/client"
	"github.com/cmsgov/easi-app/pkg/flags"
)

const (
	cedarCoreEnabledKey     = "cedarCoreEnabled"
	cedarCoreEnabledDefault = false
)

// NewClient builds the type that holds a connection to the CEDAR Core API
func NewClient(ctx context.Context, cedarHost string, cedarAPIKey string, ldClient *ld.LDClient) *Client {
	fnEmit := func(ctx context.Context) bool {
		lduser := flags.Principal(ctx)
		result, err := ldClient.BoolVariation(cedarCoreEnabledKey, lduser, cedarCoreEnabledDefault)
		if err != nil {
			appcontext.ZLogger(ctx).Info(
				"problem evaluating feature flag",
				zap.Error(err),
				zap.String("flagName", cedarCoreEnabledKey),
				zap.Bool("flagDefault", cedarCoreEnabledDefault),
				zap.Bool("flagResult", result),
			)
		}
		return result
	}

	c := cache.New(cache.NoExpiration, cache.NoExpiration) // Don't expire data _or_ clean it up

	hc := http.DefaultClient

	client := &Client{
		cedarCoreEnabled: fnEmit,
		auth: httptransport.APIKeyAuth(
			"x-Gateway-APIKey",
			"header",
			cedarAPIKey,
		),
		sdk: apiclient.New(
			httptransport.New(
				cedarHost,
				apiclient.DefaultBasePath,
				apiclient.DefaultSchemes,
			),
			strfmt.Default,
		),
		hc:    hc,
		cache: c,
	}

	// Start cache refresh for systems
	client.startCacheRefresh(ctx, time.Minute*5, client.populateSystemSummaryCache)

	return client
}

// Client represents a connection to the CEDAR Core API
type Client struct {
	cedarCoreEnabled func(context.Context) bool
	auth             runtime.ClientAuthInfoWriter
	sdk              *apiclient.CEDARCoreAPI
	hc               *http.Client
	cache            *cache.Cache
}

// startCacheRefresh starts a goroutine that will run `populateCache` based on cacheRefreshTime
// This function returns no errors, and only logs when something goes wrong
func (c *Client) startCacheRefresh(ctx context.Context, cacheRefreshTime time.Duration, populateCache func(context.Context) error) {
	ticker := time.NewTicker(cacheRefreshTime)
	go func(ctx context.Context) {
		for {
			err := populateCache(ctx)
			if err != nil {
				appcontext.ZLogger(ctx).Error("Failed to refresh cache", zap.Error(err))
			}
			// Wait for the ticker. This will block the current goroutine until the ticker sends a message over the channel
			<-ticker.C
		}
	}(ctx)
}
