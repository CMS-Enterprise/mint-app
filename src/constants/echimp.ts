export const ECHIMP_URL_SSO = `${import.meta.env.VITE_ECHIMP_URL}/eeChimpWeb/sso`;

export default ECHIMP_URL_SSO;
