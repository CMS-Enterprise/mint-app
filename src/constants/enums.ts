const TaskListSectionKeys: string[] = [
  'basics',
  'generalCharacteristics',
  'participantsAndProviders',
  'beneficiaries',
  'opsEvalAndLearning',
  'payments',
  'operationalNeeds'
];

export default TaskListSectionKeys;
