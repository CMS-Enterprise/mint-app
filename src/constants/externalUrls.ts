export const REPORT_PROBLEM_ACCESSIBILITY_TEAM_SURVEY =
  'https://docs.google.com/forms/d/e/1FAIpQLSfRJIddKkcffFWXB4AVG_ibYgShu1t-HcKuXov06mZDDxwjLw/viewform';
export const REPORT_PROBLEM_BASIC_USER_SURVEY =
  'https://www.surveymonkey.com/r/GCYMVY8';
