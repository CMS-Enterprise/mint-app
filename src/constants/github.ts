export const owner: string = 'cms-enterprise';
export const repo: string = 'mint-app';
export const filePath: string = 'src/i18n/en-US/helpAndKnowledge/Articles/';
