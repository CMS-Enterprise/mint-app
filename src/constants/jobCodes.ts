export const BASIC_USER_PROD = 'EASI_P_USER';
export const GOVTEAM_DEV = 'EASI_D_GOVTEAM';
export const GOVTEAM_PROD = 'EASI_P_GOVTEAM';
export const ACCESSIBILITY_TESTER_DEV = 'EASI_D_508_TESTER';
export const ACCESSIBILITY_TESTER_PROD = 'EASI_P_508_TESTER';
export const ACCESSIBILITY_ADMIN_DEV = 'EASI_D_508_USER';
export const ACCESSIBILITY_ADMIN_PROD = 'EASI_P_508_USER';

export const JOB_CODES = [
  BASIC_USER_PROD,
  GOVTEAM_DEV,
  GOVTEAM_PROD,
  ACCESSIBILITY_TESTER_DEV,
  ACCESSIBILITY_TESTER_PROD,
  ACCESSIBILITY_ADMIN_DEV,
  ACCESSIBILITY_ADMIN_PROD
];

export default JOB_CODES;
