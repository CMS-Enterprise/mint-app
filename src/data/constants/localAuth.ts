// eslint-disable-next-line import/prefer-default-export
export const localAuthStorageKey = 'dev-user-config';
