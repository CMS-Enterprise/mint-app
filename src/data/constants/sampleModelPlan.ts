const SAMPLE_MODEL_UUID_STRING = 'f25d8f70-6470-47e6-a6d9-debc10f26567';

export default SAMPLE_MODEL_UUID_STRING;
