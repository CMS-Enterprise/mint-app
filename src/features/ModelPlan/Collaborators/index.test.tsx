import React from 'react';
import { Provider } from 'react-redux';
import { createMemoryRouter, RouterProvider } from 'react-router-dom';
import { MockedProvider } from '@apollo/client/testing';
import {
  render,
  screen,
  waitFor,
  waitForElementToBeRemoved
} from '@testing-library/react';
import {
  GetModelCollaboratorsDocument,
  GetModelCollaboratorsQuery,
  TeamRole
} from 'gql/generated/graphql';
import configureMockStore from 'redux-mock-store';

import { ASSESSMENT } from 'constants/jobCodes';
import MessageProvider from 'contexts/MessageContext';

import { CollaboratorsContent } from './index';

type GetModelPlanType = GetModelCollaboratorsQuery['modelPlan'];

const mockAuthReducer = {
  isUserSet: true,
  groups: [ASSESSMENT],
  euaId: 'ABCD'
};

const mockStore = configureMockStore();
const store = mockStore({ auth: mockAuthReducer });

const mockCollaborator: GetModelPlanType = {
  __typename: 'ModelPlan',
  id: '123',
  modelName: 'My Model',
  collaborators: [
    {
      __typename: 'PlanCollaborator',
      userID: '123',
      id: '61c7b30c-969d-4dd4-b13b-a5065f43be43',
      modelPlanID: 'ce3405a0-3399-4e3a-88d7-3cfc613d2905',
      userAccount: {
        id: '890',
        __typename: 'UserAccount',
        email: '',
        username: 'ABCD',
        commonName: 'John Doe'
      },
      teamRoles: [TeamRole.MODEL_LEAD],
      createdDts: '2022-10-22T00:00:00Z'
    }
  ]
};

const mocks = [
  {
    request: {
      query: GetModelCollaboratorsDocument,
      variables: { id: 'ce3405a0-3399-4e3a-88d7-3cfc613d2905' }
    },
    result: {
      data: {
        modelPlan: mockCollaborator
      }
    }
  }
];

describe('Collaborator/Team Member page w/table', () => {
  it('displays a table with collaborator', async () => {
    const router = createMemoryRouter(
      [
        {
          path: '/models/:modelID/collaboration-area/collaborators',
          element: (
            <MessageProvider>
              <CollaboratorsContent />
            </MessageProvider>
          )
        }
      ],
      {
        initialEntries: [
          '/models/ce3405a0-3399-4e3a-88d7-3cfc613d2905/collaboration-area/collaborators?view=add'
        ]
      }
    );

    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Provider store={store}>
          <RouterProvider router={router} />
        </Provider>
      </MockedProvider>
    );

    expect(await screen.findByRole('table')).toBeInTheDocument();

    await waitFor(() => {
      expect(screen.getByText('Model Lead')).toBeInTheDocument();
    });

    await waitFor(() => {
      expect(screen.getByText('John Doe')).toBeInTheDocument();
    });
  });

  it('matches snapshot', async () => {
    const router = createMemoryRouter(
      [
        {
          path: '/models/:modelID/collaboration-area/collaborators',
          element: (
            <MessageProvider>
              <CollaboratorsContent />
            </MessageProvider>
          )
        }
      ],
      {
        initialEntries: [
          '/models/ce3405a0-3399-4e3a-88d7-3cfc613d2905/collaboration-area/collaborators?view=add'
        ]
      }
    );

    const { asFragment } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Provider store={store}>
          <RouterProvider router={router} />
        </Provider>
      </MockedProvider>
    );

    await waitForElementToBeRemoved(() => screen.getByTestId('page-loading'));

    expect(asFragment()).toMatchSnapshot();
  });
});
