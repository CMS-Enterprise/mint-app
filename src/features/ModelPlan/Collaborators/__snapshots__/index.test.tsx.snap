// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Collaborator/Team Member page w/table > matches snapshot 1`] = `
<DocumentFragment>
  <main
    class="mint-main-content"
    id="main-content"
    role="main"
    tabindex="-1"
  >
    <div
      class="grid-container"
      data-testid="gridContainer"
    >
      <div
        class="grid-row grid-gap"
        data-testid="grid"
      >
        <div
          class="desktop:grid-col-12"
          data-testid="grid"
        >
          <nav
            aria-label="Breadcrumbs"
            class="usa-breadcrumb usa-breadcrumb--wrap"
          >
            <ol
              class="usa-breadcrumb__list"
            >
              <li
                class="usa-breadcrumb__list-item"
              >
                <a
                  class="usa-link usa-breadcrumb__link"
                  data-discover="true"
                  href="/"
                >
                  <span>
                    Home
                  </span>
                </a>
              </li>
              <li
                aria-current="page"
                class="usa-breadcrumb__list-item usa-current"
              >
                <span>
                  Manage model team
                </span>
              </li>
            </ol>
          </nav>
          <h1
            aria-live="polite"
            class="mint-h1 margin-top-4 margin-bottom-2"
            tabindex="-1"
          >
            Add model team members
          </h1>
          <div
            class="font-body-md margin-bottom-4"
          >
            Add any team members who will be collaborating on your Model Plan. You can edit this list at any time.
          </div>
          <h4
            class="margin-bottom-1 margin-top-4"
          >
            Team members
          </h4>
          <a
            class="usa-button margin-bottom-0"
            data-discover="true"
            href="/models/ce3405a0-3399-4e3a-88d7-3cfc613d2905/collaboration-area/collaborators/add-collaborator?view=add"
          >
            Add team member
          </a>
          <div
            class="collaborator-table"
          >
            <table
              class="usa-table usa-table--borderless width-full"
              data-testid="table"
            >
              <caption
                class="usa-sr-only"
              >
                requestsTable.caption
              </caption>
              <thead>
                <tr
                  role="row"
                >
                  <th
                    aria-sort="none"
                    class="table-header"
                    colspan="1"
                    role="columnheader"
                    scope="col"
                    style="padding-bottom: .5rem; width: 25%;"
                  >
                    <button
                      class="usa-button usa-button--unstyled position-relative"
                      style="cursor: pointer;"
                      title="Toggle SortBy"
                      type="button"
                    >
                      Name
                      <svg
                        aria-label="unfold"
                        class="usa-icon margin-left-05"
                        data-testid="caret--sort"
                        focusable="false"
                        height="1em"
                        role="img"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M12 5.83 15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9 12 5.83zm0 12.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15 12 18.17z"
                        />
                      </svg>
                    </button>
                  </th>
                  <th
                    aria-sort="none"
                    class="table-header"
                    colspan="1"
                    role="columnheader"
                    scope="col"
                    style="padding-bottom: .5rem; width: 45%;"
                  >
                    <button
                      class="usa-button usa-button--unstyled position-relative"
                      style="cursor: pointer;"
                      title="Toggle SortBy"
                      type="button"
                    >
                      Role(s)
                      <svg
                        aria-label="unfold"
                        class="usa-icon margin-left-05"
                        data-testid="caret--sort"
                        focusable="false"
                        height="1em"
                        role="img"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M12 5.83 15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9 12 5.83zm0 12.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15 12 18.17z"
                        />
                      </svg>
                    </button>
                  </th>
                  <th
                    aria-sort="none"
                    class="table-header"
                    colspan="1"
                    role="columnheader"
                    scope="col"
                    style="padding-bottom: .5rem; width: auto;"
                  >
                    <button
                      class="usa-button usa-button--unstyled position-relative"
                      style="cursor: pointer;"
                      title="Toggle SortBy"
                      type="button"
                    >
                      Date added
                      <svg
                        aria-label="unfold"
                        class="usa-icon margin-left-05"
                        data-testid="caret--sort"
                        focusable="false"
                        height="1em"
                        role="img"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M12 5.83 15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9 12 5.83zm0 12.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15 12 18.17z"
                        />
                      </svg>
                    </button>
                  </th>
                  <th
                    aria-sort="none"
                    class="table-header"
                    colspan="1"
                    role="columnheader"
                    scope="col"
                    style="padding-bottom: .5rem; width: auto;"
                  >
                    <button
                      class="usa-button usa-button--unstyled position-relative"
                      type="button"
                    >
                      Actions
                      <svg
                        aria-label="unfold"
                        class="usa-icon margin-left-05"
                        data-testid="caret--sort"
                        focusable="false"
                        height="1em"
                        role="img"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M12 5.83 15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9 12 5.83zm0 12.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15 12 18.17z"
                        />
                      </svg>
                    </button>
                  </th>
                </tr>
              </thead>
              <tbody
                role="rowgroup"
              >
                <tr
                  role="row"
                >
                  <td
                    role="cell"
                    style="white-space: normal;"
                  >
                    <div
                      class="display-flex  flex-align-center"
                    >
                      <div
                        class="display-flex flex-align-center flex-justify-center minw-4 circle-4 bg-blue-10 margin-right-1"
                        data-testid="avatar--basic"
                      >
                        JD
                      </div>
                      <div
                        class="margin-y-0"
                      >
                        <span
                          class=""
                        >
                          John Doe
                        </span>
                      </div>
                    </div>
                  </td>
                  <td
                    role="cell"
                    style="white-space: normal;"
                  >
                    Model Lead
                  </td>
                  <td
                    role="cell"
                    style="white-space: normal;"
                  >
                    October 22, 2022
                  </td>
                  <td
                    role="cell"
                    style="white-space: normal;"
                  >
                    <a
                      aria-label="Edit John Doe"
                      class="usa-link margin-right-2"
                      data-discover="true"
                      href="/models/ce3405a0-3399-4e3a-88d7-3cfc613d2905/collaboration-area/collaborators/add-collaborator/61c7b30c-969d-4dd4-b13b-a5065f43be43?view=add"
                    >
                      Edit
                    </a>
                  </td>
                </tr>
              </tbody>
            </table>
            <div
              class="grid-row grid-gap grid-gap-lg display-flex row-reverse"
            />
            <div
              aria-live="polite"
              class="usa-sr-only usa-table__announcement-region"
            >
              Requests table reset to default sort order
            </div>
          </div>
          <div
            class="margin-top-5 display-block"
          >
            <a
              class="usa-button usa-button--outline"
              data-discover="true"
              data-testid="button--back"
              href="/models/new-plan"
            >
              Back
            </a>
            <a
              class="usa-button usa-button--outline"
              data-discover="true"
              data-testid="continue-to-collaboration-area"
              href="/models/ce3405a0-3399-4e3a-88d7-3cfc613d2905/collaboration-area"
            >
              Continue to model collaboration area
            </a>
          </div>
        </div>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
